### ISSUE TITLE
Upgrade to 7.1 stuck "Fabric upgrade failed on the node"

### ISSUE THREAD

**Issue by julioas09** [microsoft/service-fabric#955]
Long story short, a customer as 4 standalone clusters (environments).
They upgraded 2 environments from 6.5.664.9590 to 7.1.490.9590, and in one of them they had to copy the 7.1 .cab manually on one node because the 'download failed on the node. User or password is wrong' and it completed successfully after that.

After that, they were recommended to upgrade first to 7.0CU3 and then to 7.1 to avoid that for the other 2 environments. They followed that approach and in the next environment (standalone 7 node WS2016 cluster) they upgraded to 7.0CU3, all fine.
Then they triggered the upgrade to 7.1 and on the first node, the upgrade got stuck "Fabric upgrade failed on the node".

![as1](https://user-images.githubusercontent.com/3543529/84141528-9c544f80-aa53-11ea-95ee-0ca272f23b99.png)

![as2](https://user-images.githubusercontent.com/3543529/84141534-9eb6a980-aa53-11ea-8f2d-7342dcaada8e.png)

The node shows this Errors on the events:

`FinishTryDownloadCodePackage: FabricCodePackage download failed. ErrorCode=FABRIC_E_CANNOT_CONNECT StoreFile=WindowsFabricStore\WindowsFabric.7.1.409.9590.cab LocalFile=E:\SF\Data\ibsfabfe01prup\Fabric\work\FabricUpgrade\WindowsFabric.7.1.409.9590.cab`
![as5](https://user-images.githubusercontent.com/3543529/84141729-ee957080-aa53-11ea-9466-777a410e4d79.png)

`-ibsfabfe01prup.xxxxxxxxxx:19000: connection auth status received for 28b261c6840: FABRIC_E_CONNECTION_DENIED:FABRIC_E_CONNECTION_DENIED: E_FAIL, role granted: User`
![as7](https://user-images.githubusercontent.com/3543529/84141742-f228f780-aa53-11ea-9ee0-bf68ed5c25af.png)

![as8](https://user-images.githubusercontent.com/3543529/84141756-f523e800-aa53-11ea-933a-a383223342b1.png)

The file is actually on that path on the node:
![as6](https://user-images.githubusercontent.com/3543529/84139929-0f0ffb80-aa51-11ea-9421-3c7df9130764.png)

Any recommendation on how to proceed? Any idea on what could be the cause?

**Comment by maburlik** [comment]
Thanks for your due diligence in discovering the issue. The cause is there was a bit flip in the CAB that was delivered and you are finding that the file integrity has been corrupted. This could have been due to an interrupted download (unlikely) or a rare case that we've seen where the download center cache temporarily corrupts, and in cases we've found this the issue self-corrects within a short period of time.

The solution is to roll back any pending upgrade via:
Update-ServiceFabricClusterUpgrade -UpgradeDomainTimeoutSec 10 -UpgradeTimeoutSec 20 -HealthCheckRetryTimeoutSec 1 -HealthCheckWaitDurationSec 1 -HealthCheckStableDurationSec 1

Unprovision the code package using:
Unregister-ServiceFabricClusterPackage https://docs.microsoft.com/en-us/powershell/module/ServiceFabric/Unregister-ServiceFabricClusterPackage?view=azureservicefabricps

Then if your cluster has automatic upgrade enabled it will update automatically within 24 hours, otherwise download the CAB manually using the link retrieved by calling the command:
Get-ServiceFabricRuntimeUpgradeVersion -BaseVersion 7.0.466.9590

and run as detailed on this page: https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-upgrade-windows-server
Copy-ServiceFabricClusterPackage https://docs.microsoft.com/en-us/powershell/module/servicefabric/copy-servicefabricclusterpackage?view=azureservicefabricps
Register-ServiceFabricClusterPackage https://docs.microsoft.com/en-us/powershell/module/servicefabric/register-servicefabricclusterpackage?view=azureservicefabricps
Start-ServiceFabricClusterUpgrade https://docs.microsoft.com/en-us/powershell/module/servicefabric/start-servicefabricclusterupgrade?view=azureservicefabricps

**Comment by julioas09** [comment]
We noticed that when trying to open the .cab it was corrupted. What we did is copy the .cab manually from other environment to the nodes where it got stuck. This progressed until the cluster was upgraded.
The environment is divided into two nodetypes in different subnets: Frontend and Backend.
**We only had to apply this workaround in the Frontend nodes**... which may indicate some communication problems, but the primary replica of the Image Store Service was reachable from the problematic nodes.

What we see now even though the cluster is upgraded, is that when trying to deploy a new version of an application to the Frontend nodes:

![a](https://user-images.githubusercontent.com/3543529/85117883-d544af00-b21f-11ea-90d1-1e4e448d40d3.png)

![b](https://user-images.githubusercontent.com/3543529/85117896-d970cc80-b21f-11ea-91d3-ef576db33a3a.png)

The upgrade does not go through, it gets stuck, and it seems closely related to the problem faced during the cluster upgrade. :(
